Add tests for date module

oscargus opened the pull request:
## PR Summary

~Extended `date_ticker_factory` to return `SecondLocator` and `Microsecondlocator`. The format strings can be discussed.~

Removed dead code.

Added tests for more cases in `dates`.

Should we remove the special handling for dateutil <= 2.3? 2.4 appears to have been release 2014-2015 or so.

https://github.com/matplotlib/matplotlib/blob/65e962008a1f43fca4b1096c6f88685892d1ab2d/lib/matplotlib/dates.py#L1526-L1528

## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->
**Tests and Styling**
- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [N/A] New features are documented, with examples if plot related.
- [N/A] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [N/A] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [N/A] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of main, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

anntzer commented:
FWIW we already require dateutil>=2.7 (in setup.py).

oscargus commented:
Then the only problem remaining is that none of the special handling is actually tested, so hard to say if something breaks when removed...

jklymak reviewed (commented):
I'm pretty confused by this.  So far as I can tell we don't call `date_ticker_factory` in the code base.  Am I missing some magic somewhere?  Its always better to test the user facing interface, rather than the helper (which should probably be private).    So does this actually do anything for the user?  If so, what is the example?

oscargus commented:
I was under the impression that it is used to return a suitable set of locator and formatter if one inputs a range in days. It is 17 years old, so I assume that someone uses it.

But you are right that it is not used in the code base nor is it in the documentation.

Personally, I stumbled upon it while trying to increase test coverage.

oscargus commented:
At least someone uses it: https://github.com/daq-tools/kotori

https://github.com/daq-tools/kotori/blob/d25565920429c17ca4b3b28f22d4d7b160405c0a/kotori/io/export/util.py#L61-L62

QuLogic commented:
`date_ticker_factory` was replaced by `AutoDateFormatter`/`AutoDateLocator` in ... 2006: 5f9468de5597a6bec8df63d652704a188d43d52c

oscargus commented:
Sure, I can deprecate it then.

oscargus commented:
Removed the changes to `date_ticker_factory` here and will deprecate in another PR to keep it separate. Now: only added tests, dead code removal (at least I cannot trigger that part of the code and can probably argue for why it is not possible anymore), and corrected a comment.

greglucas commented:
Needs a rebase, but otherwise looks good.

oscargus commented:
Rebased.

oscargus commented:
Gah! Rebased the wrong version... Will continue when on the computer with the latest version...

oscargus commented:
Rebased the correct version...
